File: Data Structures & Algorithms/permutation-string/submission-0.py
class Solution:
    def checkInclusion(self, s: str, t: str) -> bool:
        mapS = {}
        mapT = {}
        left = 0
        for ch in s:
            mapS[ch] = mapS.get(ch,0) + 1
        
        for i in range(len(t)):
            mapT[t[i]] = mapT.get(t[i],0) + 1
            while (i - left + 1 > len(s)):
                mapT[t[left]] = mapT.get(t[left],0) - 1
                if mapT[t[left]] == 0:
                    del mapT[t[left]]
                left += 1
            if mapS == mapT:
                return True
        return False



        

File: Data Structures & Algorithms/minimum-window-with-characters/submission-0.swift
class Solution {
    func minWindow(_ s: String, _ t: String) -> String {
        var mapS: [Character: Int] = [:]
        var mapT: [Character: Int] = [:]
        var minlenth = Int.max
        var left = 0
        var start = 0
        for ch in t {
            mapT[ch, default: 0] += 1
        }
        let arr = Array(s)
        for i in 0..<arr.count {
            mapS[arr[i], default:0] += 1
            while(contains(mapS,mapT)) {
                if i - left + 1 < minlenth {
                    minlenth = i - left + 1
                    start = left
                }
                mapS[arr[left], default:0] -= 1
                if mapS[arr[left]] == 0 {
                    mapS.removeValue(forKey: arr[left])
                }
                left += 1
            }
        }
        if minlenth == Int.max { return "" }
        return String(arr[start..<start + minlenth])
    }

    func contains( _ mapS: [Character:Int], _ mapT: [Character:Int]) -> Bool {
        for (key,value) in mapT {
            if mapS[key, default: 0] < value {
                return false
            }
        }
        return true
    }
}
